### ISSUE TITLE
Asian name inverted in 'name' attribute

### ISSUE THREAD

**Issue by hiroshisuga** [keycloak/keycloak#31271]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

user-profile

### Describe the bug

The external authentication using Microsoft Entra ID generates the 'name' attribute inverted, namely, firstname - familyname order, in Asian names. In some Asian countries and others, a name is in familyname - firstname order, so the current implementation looks quite strange to us.
In the source code
https://github.com/keycloak/keycloak/blob/c33585a5f45322e2bd0a06465f72e4bf5b161e6c/server-spi-private/src/main/java/org/keycloak/broker/provider/BrokeredIdentityContext.java#L334
, Keycloak appears to generate 'name' in firstname - familyname order when 'name' property is not provided by the identity provider, which seems to be the case for Microsoft EntraID (at least in my domain). Google provides 'name' property correctly.
I would like to propose to use displayName property of Microsoft EntraID to generate the 'name' attribute. 

### Version

20.0.5 (sorry, this is only available in my environment)

### Regression

- [ ] The issue is a regression

### Expected behavior

Asian names are generated in family name - first name order, when written in Chinese letter. But when written in alphabet, we are fine with first name - family name order. So the simplest way is to use displayName (when available) when the 'name' property is not provided by the Idp.

### Actual behavior

All names are always generated in first name - family name order, irrespective of locale and used letters. 

### How to Reproduce?

I can confirm in the log. 

### Anything else?

A related post is here: 
https://github.com/bigbluebutton/greenlight/issues/5814
, which is the first problem that I encountered.

**Comment by pedroigor** [comment]
@hiroshisuga Thanks. Can you please elaborate what you mean by "I would like to propose to use displayName property of Microsoft EntraID to generate the 'name' attribute.". 

Just to confirm my understanding, the IdP is returning the `name` where the last name (family name) is at the beginning. That cause, the last name to be set to the `first name` and vice-versa. Is that the case?

I'm not sure how we can safely set those names without introducing some setting to change how we parse the names when the last happens to be the first. For instance, we could have `Last name first` switch at the broker config. I'm not sure if we can rely on the user locale (and how to safely resolve it during authentication).

Do you know if EntraID provides any hint about the name format so we can use it to change how we parse the name?

**Comment by pedroigor** [comment]
Now I see what you mean by using the displayName. The StackOverflow broker is already using a separated field (also called display name) to fetch the user name.

Are you able to give a link to the EntraID documentation where they document the value of the `displayName`?

@sschu Do you think it is safe to use the `displayName` from EntraID responses/tokens to resolve the user's name and parse it to both first and last name attributes?

**Comment by hiroshisuga** [comment]
@pedroigor , thanks for the answer.
I do not really find the detailed explanation how displayName is set. It seems what the organisations decide..., so I guess it is what the users want to use for showing their names.
In my organisation, displayName is the one that is used for displaying the name in all MS apps such as MS Teams. 
I found only this info:
https://learn.microsoft.com/en-us/windows/win32/adschema/a-displayname

**Comment by pedroigor** [comment]
Thank you @hiroshisuga. By looking at https://learn.microsoft.com/en-us/entra/identity-platform/id-token-claims-reference, I don't see any claims related to "displayName".

However, some brokers we have do use a `display_name` claim to resolve the full name (and eventually setting the first and last name). I'm wondering if you can't configure EntraID to set a specific value in the `name` claim. That is the claim we  use to set the first and last names if they are not available from the token issued by EntraID.

Ohter than that, I'm not sure if we can do much but you can always extend our built-in implementation to fulfill your needs, if you can't sort it out when configuring EntraID.

The better solution would be to eventually allow reading the full name (e.g.: display name you mentioned) from a claim other than `name`, which is hard coded at the moment. For that, we need to at least know you can set a different claim to tokens issued by EntraID with the format you want.

**Comment by hiroshisuga** [comment]
Thanks @pedroigor . I understand what you say.
Changing the organisation's EntraID setting is not allowed to me. So I am trying to map the displayName attribute into the 'name' attribute by the Keycloak setting. So what I did is:
1. Identity providers -> microsoft (I created) -> Mappers -> map MS 'displayName' to user attribute 'displayName'
2. Clients -> greenlight (I created) -> Client scopes -> greenlight-dedicated -> mappers -> map the user attribute 'displayName' to token claim 'displayName'

With those setting, I can refer the claim 'displayName' from the application (Greenlight). However, I need to modify the source code of Greenlight which natively fetch the 'name' claim.

Would it be possible to rename 'displayName' to 'name' so that Greenlight app can fetch the 'name' claim without changing the source code? I changed the Token Claim Name in the user attribute mapper but it did not work out.
